src/python/grapl_analyzerlib/grapl_analyzerlib/nodes/asset.py

```diff
@@ -0,0 +1,213 @@
+from grapl_analyzerlib.viewable import Viewable, V, Q
+from grapl_analyzerlib.comparators import StrCmp, Eq, Distance
+
+AQ = TypeVar("AQ", bound="AssetQuery")
+AV = TypeVar("AV", bound="AssetView")
+
+
+def default_asset_properties() -> Dict[str, PropType]:
+    return {
+        "hostname": PropType(PropPrimitive.Str, False,),
+    }
+
+
+def default_asset_edges() -> Dict[str, Tuple[EdgeT, str]]:
+    return {
+        "asset_ip": (
+            EdgeT(AssetSchema, IpAddressSchema, EdgeRelationship.OneToMany),
+            "ip_assigned_to",
+        ),
+        "asset_processes": (
+            EdgeT(AssetSchema, ProcessSchema, EdgeRelationship.OneToMany,),
+            "process_asset",
+        ),
+        "files_on_asset": (
+            EdgeT(AssetSchema, FileSchema, EdgeRelationship.OneToMany,),
+            "file_asset",
+        ),
+    }
+
+
+class AssetSchema(Schema):
+    def __init__(self):
+        super(AssetSchema, self).__init__(
+            default_asset_properties(), default_asset_edges(), lambda: AssetView
+        )
+
+    @staticmethod
+    def self_type() -> str:
+        return "Asset"
+
+
+class AssetQuery(Queryable[AV, AQ]):
+    @classmethod
+    def node_schema(cls) -> "Schema":
+        return AssetSchema()
+
+    def with_hostname(
+        self,
+        *,
+        eq: Optional["StrOrNot"] = None,
+        contains: Optional["OneOrMany[StrOrNot]"] = None,
+        starts_with: Optional["StrOrNot"] = None,
+        ends_with: Optional["StrOrNot"] = None,
+        regexp: Optional["OneOrMany[StrOrNot]"] = None,
+        distance_lt: Optional[Tuple[str, int]] = None,
+    ) -> "AssetQuery":
+        self._property_filters["hostname"].extend(
+            _str_cmps(
+                predicate="hostname",
+                eq=eq,
+                contains=contains,
+                ends_with=ends_with,
+                starts_with=starts_with,
+                regexp=regexp,
+                distance_lt=distance_lt,
+            )
+        )
+        return self
+
+    def with_asset_ip(self, *asset_ips: 'IpAddressSchema'):
+        asset_ips = asset_ips or [IpAddressSchema()]
+        self.set_neighbor_filters("asset_ip", [asset_ips])
+        for asset_ip in asset_ips:
+            asset_ip.set_neighbor_filters("ip_assigned_to", [self])
+
+    def with_asset_processes(self, *asset_processes: 'ProcessSchema'):
+        asset_processes = asset_processes or [ProcessSchema()]
+        self.set_neighbor_filters("asset_processes", [asset_ip])
+        for asset_process in asset_processes:
+            asset_process.set_neighbor_filters("process_asset", [self])
+
+    def with_files_on_asset(self, *files_on_asset: 'FileSchema'):
+        files_on_asset = files_on_asset or [FileSchema()]
+        self.set_neighbor_filters("files_on_asset", [asset_ip])
+        for file_on_asset in files_on_asset:
+            file_on_asset.set_neighbor_filters("file_asset", [self])
+
+    @staticmethod
+    def extend_self(*types):
```

Comment:
Devils advocate:
It looks like every `extend_self` is basically the same except with swapping out the class and the class name. Could we make a
```
    @classmethod
    def extend_self(cls, *types):
      return extend_self_helper(cls, *types)

def extend_self_helper(cls, *types):
          cls_name = cls.__name__  # this might need to have the module name stripped, i forget
          for t in types:
            method_list = [
                method for method in dir(t) if method.startswith("__") is False
            ]
            for method in method_list:
                setattr(cls, method, getattr(t, method))
        return type(cls_name, types, {})
```
? (may need some selftype-ing or something)

Reply:
Absolutely, I tried that and it didn't work. I want to revisit it though.